python/tests/__init__.py

```diff
@@ -20,11 +20,113 @@
+        )
+        tables = tskit.TableCollection.fromdict(tables_dict)
+        if provenance_record is not None:
+            tables.provenances.add_row(provenance_record)
+        if self.from_ts is None:
+            # Add the populations with metadata
+            assert len(tables.populations) == len(self.population_configurations)
+            tables.populations.clear()
+            for pop_config in self.population_configurations:
+                tables.populations.add_row(metadata=pop_config.encoded_metadata)
+        tables.build_index()
+        tables.compute_mutation_times()
+        return tables.tree_sequence()
+
+    def mutate(
+        tree_sequence,
+        rate=None,
+        random_seed=None,
+        model=None,
+        keep=False,
+        start_time=None,
+        end_time=None,
+    ):
+        try:
+            tables = tree_sequence.tables
+        except AttributeError:
+            raise ValueError("First argument must be a TreeSequence instance.")
+        if random_seed is None:
+            random_seed = msprime.simulations._get_random_seed()
+        random_seed = int(random_seed)
+
+        rng = msprime.mutations._msprime.RandomGenerator(random_seed)
+        if model is None:
+            model = msprime.InfiniteSites()
+        try:
+            alphabet = model.alphabet
+        except AttributeError:
+            raise TypeError("model must be an InfiniteSites instance")
+        if rate is None:
+            rate = 0
+        rate = float(rate)
+        keep = bool(keep)
+
+        parameters = {
+            "command": "mutate",
+            "rate": rate,
+            "random_seed": random_seed,
+            "keep": keep,
+        }
+
+        if start_time is None:
+            start_time = -sys.float_info.max
+        else:
+            start_time = float(start_time)
+            parameters["start_time"] = start_time
+
+        if end_time is None:
+            end_time = sys.float_info.max
+        else:
+            end_time = float(end_time)
+            parameters["end_time"] = end_time
+        provenance_dict = msprime.provenance.get_provenance_dict(parameters)
+
+        if start_time > end_time:
+            raise ValueError("start_time must be <= end_time")
+
+        mutation_generator = msprime.mutations._msprime.MutationGenerator(
+            rng, rate, alphabet=alphabet, start_time=start_time, end_time=end_time
+        )
+        lwt = msprime.mutations._msprime.LightweightTableCollection()
+        lwt.fromdict(tables.asdict())
+        mutation_generator.generate(lwt, keep=keep)
+
+        tables_dict = lwt.asdict()
+        tables_dict["mutations"]["time"] = np.full(
+            tables_dict["mutations"]["node"].shape, -1, np.float64
+        )
+        tables = tskit.TableCollection.fromdict(tables_dict)
+        tables.provenances.add_row(json.dumps(provenance_dict))
+
+        tables.build_index()
+        tables.compute_mutation_times()
```

Comment:
Is the idea that when msprime adds in times itself that this won't be needed? When that happens, a bunch of the checks that times can be exactly recomputed by compute_mutation_times will fail; maybe it's better to not assume that times coming in from mutate are regularly spaced?

Reply:
Yeah, I copied this test structure from mutation parents. I thought mutation times might be configurable in msprime and that this test could specify what it wanted. But this this test isn't doing much meaningful testing though.

Reply:
I'm not sure, but my guess is that adding mutations regularly along branches would be much harder in msprime than putting them uniformly (since you'd have to know the future to do regularly spaced). And, I can't think of a reason anyone would want that.

Reply:
Tests now don't use msprime times, only msprime topology in https://github.com/tskit-dev/tskit/pull/672/commits/93e65c6ddc6f4768c1ac07576e62979bb5a04604